### ISSUE TITLE
module loaders

### ISSUE THREAD

**Issue by stagas** [isojs/isojs#2]
`npm`, `bower`, `component`, `duo`, `jspm`, `browserify`, `systemjs`, `webpack`, `meteor`. I'm sure there are more. Everyone is doing it differently. This is by far the biggest pain currently in JS land.

**Comment by stagas** [comment]
The idea is, if you spend more time writing configuration files, makefiles, manifest.jsons, publishing, etc. than the time you spend to actually write the code, then something is wrong with that process. Yes, you can have templates and cli tools to aid with that, but still, that's a workaround. The issue is still there.

**Comment by yields** [comment]
i agree, but: https://xkcd.com/927/

**Comment by stagas** [comment]
I'm with you on this, that's why I suggest we use existing standards rather than invent new ones.

My proposal consists of two simple things:

- A module is a resource, therefore It makes sense that all import statements are specified as URLs and are deterministic in nature:

```js
import merge from 'https://cdn.rawgit.com/yields/merge/1.0.0/index.js';
```

- Basic metadata can live along the source code and defined at the top of the entry file:

```js
/**
 * @module merge
 * @author yields
 * @version 1.0.0
 * @repo https://github.com/yields/merge
 * @license mit
 */
```

Pros:

- The source code is the entire thing. Less time spent in writing json manifest files, more time left for coding. It's also portable. Tutorials and blog posts can have example code that works by just a copy paste.
- Dependencies are immediately available. There is no need to define the resource location in another file than the actual source file that needs it, thus manifest files become obsolete.
- Fast lookup and retrieval. When you don't need the extra roundtrip to a registry or repository to check versions to use, installs are faster, CDNs like `rawgit` make it even faster for everyone worldwide.
- Cached forever. A certain import statement will always resolve to the same file, therefore it can be cached locally and that makes builds extremely fast.
- Deterministic. You can guarantee that the exact same build is available in every environment. If it builds and works locally, the exact same code is generated and works everywhere. Also, never again random updates that break the build caused by erroneous versioning.
- Can be understood and processed by almost every tool that deals with URLs. It can be understood by humans and inspected by just a copy paste to the browser bar. No hunting around where that dependency lives.
- A naive implementation of a loader like this is under 100 loc. All the complexity rises from parsing json files, version management and workarounds for problems introduced by that complexity itself.

Cons:
- Verbose import statements. It's a tradeoff. A bit of ugliness, for better stability and getting rid of manifests.
- No automagic updates. This is a pro for me, but usually not what people want. I believe updates should be explicit, but with help of tools we can statically replace import statements with the updated versions if necessary.
- ?

**Comment by ramitos** [comment]
@stagas have you heard of http://normalize.github.io/ ?

**Comment by stagas** [comment]
@ramitos yeah, kinda in the right direction but still introduces yet another format for import statements which relies on manifests and kind of cancels all the pros I just wrote :) I think we should just use URLs. That's the definite standard for resources and it's already used in HTML and everywhere. The analogous of what we're doing now would be like `<img src="cat-image^1.0">` and having a json file defining that `"cat-image": { repository: "https://wherever-it-lives/cat-image", types: ["jpeg", "png"], sizes: ["320x200", "640x400"] }`, and then have json files for every browser and every system because they all define image dependencies differently in their own registries. Sometimes you'd need to convert the image from one format to another on the fly because it isn't supported by your browser. And of course, you'd need multiple round trips, reading the json, checking the repository for the updated cat image, updating if necessary.

**Comment by stagas** [comment]
@ramitos sorry, I mean deterministic URLs. I see that nlz does version resolution server-side, meaning a certain URL will not always be the same file. However, I don't know, it's not relevant then to this conversation. If people want to point their dependencies to nlz, they can, I guess. They just can't cache locally and always need to hit the server for updates and kind of cancels all the deterministic build advantages I mentioned.

**Comment by stagas** [comment]
@ramitos but then again, the proper way I guess is using http cache headers for that stuff. Some might be cached forever, some not and it's fine.

**Comment by stagas** [comment]
It looks good though. I think it's in the direction we should be aiming for.

**Comment by yields** [comment]
what about older browsers? will there be a tool to transform etc..?

**Comment by stagas** [comment]
@yields maybe, perhaps. My idea is, we design for the *ideal* workflow, not something production-ready today, and build tools around that. One ideal for me would be a repo containing just an `index.js` and a `test.js`. Then think upwards, *how* can we make this happen?

For example, metadata and API documentation can live in jsdoc-style comments in the `index.js`, then a `readme` tool could parse that on-the-fly from the cli: `readme yields/merge`. The `test.js` could also contain metadata for its test runner and then export just the spec cases. A `test` tool can then parse that and run `mocha` or `jasmine` or whatever using that metadata just by `test yields/merge`. Maybe also `test --gecko yields/merge` or `test --webkit yields/merge` then run slimer or phantom. A `test --all` could run the same spec against all environments. Using some simple conventions we can agree on, we can abstract away all those details that bring friction to development.

These are just ideas to improve the workflow, keeping things tight, concise and portable. There is just too much repetitive work and friction that we can eliminate and I don't see much work happening towards that, instead we constantly add more friction that essentially work around the fact that there are no standards or conventions in userland workflow. Just as there are standards at the core level, I believe there should also be formal definitions of workflow so that people don't do chaotic decisions. I just don't enjoy the fact that more than half of the stuff people put in their repos require me to install new, different tools that do the same thing and most of the time fail because I'm on a windows machine.

**Comment by stagas** [comment]
Bottom line, we shouldn't be writing Makefiles. We should be writing code and it should work everywhere, be testable and importable with documentation by just a git push. The ideal workflow, `mkdir foo`  write an `index.js` and a `test.js` and git push it and it should be available for every environment that deals with javascript. Not eliminate the extra, but keep them optional, not mandatory.

**Comment by jonathanong** [comment]
i'm gonna wait until es6 modules are out before working more on normalize, but we could start using https://github.com/normalize/registry.js for a consistent URL naming scheme.

**Comment by stagas** [comment]
see https://github.com/babel/babel/issues/1291
